sql/1715118159.sql

```diff
@@ -0,0 +1,112 @@
+-- +migrate Up
+-- +migrate StatementBegin
+create or replace function convoy.take_token(_key text, _rate integer, _bucket_size integer) returns boolean
+    language plpgsql
+as
+$$
+DECLARE
+    next_min timestamptz;
+    can_take BOOLEAN;
+    row RECORD;
+BEGIN
+    SELECT expires_at, tokens FROM convoy.token_bucket WHERE key = _key FOR UPDATE SKIP LOCKED LIMIT 1 INTO row;
+    if row is null then
```

Comment:
@jirevwe, I'm probably missing something here, but we're unable to differentiate between if the row is locked by another transaction or if the row does not exist at all, so the `IF NOT FOUND` below will never run.

Reply:
`if row is null then` is for the select query.

`IF NOT FOUND THEN` captures the updated row from the update query. If there was no update it'll run the insert query.